worldedit-bukkit/adapters/adapter-1_21_4/src/main/java/com/sk89q/worldedit/bukkit/adapter/impl/fawe/v1_21_4/PaperweightPlatformAdapter.java

```diff
@@ -0,0 +1,724 @@
+                        LOGGER.warn("Chunk {},{} failed to load in 10 seconds in world {}. Retrying...", chunkX, chunkZ, world);
+                        // Retry chunk load
+                        chunk = (CraftChunk) serverLevel.getWorld().getChunkAtAsync(chunkX, chunkZ, true, true).get();
+                    } else {
+                        throw new UnsupportedOperationException("Cannot load chunk from unloaded world " + world + "!");
+                    }
+                }
+                addTicket(serverLevel, chunkX, chunkZ);
+                return (LevelChunk) CRAFT_CHUNK_GET_HANDLE.invoke(chunk);
+            } catch (Throwable e) {
+                e.printStackTrace();
+            }
+        }
+        return TaskManager.taskManager().sync(() -> serverLevel.getChunk(chunkX, chunkZ));
+    }
+
+    private static void addTicket(ServerLevel serverLevel, int chunkX, int chunkZ) {
+        // Ensure chunk is definitely loaded before applying a ticket
+        io.papermc.paper.util.MCUtil.MAIN_EXECUTOR.execute(() -> serverLevel
+                .getChunkSource()
+                .addRegionTicket(ChunkHolderManager.UNLOAD_COOLDOWN, new ChunkPos(chunkX, chunkZ), 0, Unit.INSTANCE));
+    }
+
+    public static ChunkHolder getPlayerChunk(ServerLevel nmsWorld, final int chunkX, final int chunkZ) {
+        ChunkMap chunkMap = nmsWorld.getChunkSource().chunkMap;
+        try {
+            return (ChunkHolder) methodGetVisibleChunk.invoke(chunkMap, ChunkPos.asLong(chunkX, chunkZ));
+        } catch (Throwable thr) {
+            throw new RuntimeException(thr);
+        }
+    }
+
+    @SuppressWarnings("deprecation")
+    public static void sendChunk(IntPair pair, ServerLevel nmsWorld, int chunkX, int chunkZ) {
+        ChunkHolder chunkHolder = getPlayerChunk(nmsWorld, chunkX, chunkZ);
+        if (chunkHolder == null) {
+            return;
+        }
+        LevelChunk levelChunk;
+        if (PaperLib.isPaper()) {
+            // getChunkAtIfLoadedImmediately is paper only
+            levelChunk = nmsWorld.getChunkSource().getChunkAtIfLoadedImmediately(chunkX, chunkZ);
+        } else {
+            levelChunk = chunkHolder.getTickingChunkFuture().getNow(ChunkHolder.UNLOADED_LEVEL_CHUNK).orElse(null);
+        }
+        if (levelChunk == null) {
+            return;
+        }
+        StampLockHolder lockHolder = new StampLockHolder();
+        NMSAdapter.beginChunkPacketSend(nmsWorld.getWorld().getName(), pair, lockHolder);
+        if (lockHolder.chunkLock == null) {
+            return;
+        }
+        MinecraftServer.getServer().execute(() -> {
+            try {
+                ChunkPos pos = levelChunk.getPos();
+                // NOTE: the ClientboundForgetLevelChunkPacket packet is required on 1.21.3
+                // as the client won't update empty -> non-empty sections properly otherwise
+                ClientboundForgetLevelChunkPacket forget = new ClientboundForgetLevelChunkPacket(pos);
```

Comment:
> Spigot works so far as well, though there are weird chunk errors (partially only visible when getting to the actual section, leaving the section makes them invisible again) and some lighting issues. No error though, so idk if that's just Spigot being Spigot.
Did you try removing this again for 1.21.4? This *could* affect the behavior, and Mojang had some changes in that area in 1.21.4 that might make this workaround unnecessary again.

Reply:
Removing the forget packet doesn't really change anything in that case - no issues on paper though. Will try to investigate

https://github.com/user-attachments/assets/5724e7ca-bcd4-457e-91c9-fa9c35ee557a

Reply:
It appears to only happen 1.21.2/3 onwards - i don't have these weird glitches in 1.21.1 (but it's already present when checking in 1.21.3 Spigot using a vanilla client)